EECO-1761 Fixed Nullable object types in MS.

sebastiantorrescl opened the pull request:
It keeps the current logic of add a ` | nil ` for the non-object types, but for the objects, it places in an `anyOf:` data type, which can be the entire object or `type: nil`.

This logic is approved for Mulesoft, and it was tested in the conflicted scenario.

Jira ticket: [https://checkr.atlassian.net/browse/EECO-1761](https://checkr.atlassian.net/browse/EECO-1761)

akshay-8d66 commented:
@sebastiantorrescl Can you also posted the generated RAML. If possible only the diff.

sebastiantorrescl commented:
Sure @akshay-8d66, this is the only different scenario:
Before:

```
      record:
        description: >
          Simplified representation of linked record.

          If `null`, Candidate Story is considered general information about the
          Candidate.
        type: object | nil
        properties:
          case_number:
            description: 'Case number, if existing'
            example: NMND78D
            type: string
            (nullable): true
          charge:
            description: 'Charge description, if existing'
            example: LIGHT OFFENSE
            type: string
            (nullable): true
          id:
            description: Record ID
            example: 4e2a551f79e0095b35d4d3ca
            type: string
          location:
            description: >-
              Location of the Record, can be a State, or a City and a State, if
              existing
            example: 'San Mateo, CA'
            type: string
            (nullable): true
          offense_date:
            description: 'Offense date, if existing'
            example: '2020-01-20'
            type: date-only
            (nullable): true

```


After:
```
      record:
        anyOf:
          - description: >
              Simplified representation of linked record.

              If `null`, Candidate Story is considered general information about
              the Candidate.
            type: object
            properties:
              case_number:
                description: 'Case number, if existing'
                example: NMND78D
                type: string | nil
              charge:
                description: 'Charge description, if existing'
                example: LIGHT OFFENSE
                type: string | nil
              id:
                description: Record ID
                example: 4e2a551f79e0095b35d4d3ca
                type: string
              location:
                description: >-
                  Location of the Record, can be a State, or a City and a State,
                  if existing
                example: 'San Mateo, CA'
                type: string | nil
              offense_date:
                description: 'Offense date, if existing'
                example: '2020-01-20'
                type: date-only | nil
          - type: nil
```

Previously it didn't fix the inner nullable properties. Now it can fix even an inner object.

akshay-8d66 commented:
Looks good to me. I'll let defer the final review to @admytren if he wants something else.

admytren commented on lib/raml-fixers/nullable-fixer.js:
```diff
@@ -21,14 +27,40 @@ function customizer (input) {
+  //If we are in this point it's because we are the input it's nullable.
+  var output = {};
+  if ('type' in input) { 
+
```
Recommend deleting whitespace

admytren commented on lib/raml-fixers/nullable-fixer.js:
```diff
@@ -21,14 +27,40 @@ function customizer (input) {
+    //Nullable without a type.
+    output = _.cloneDeep(input);
   }
   deleteNullable(output);
```
Recommend moving deleteNullable into scope on line 44 and 48, otherwise it's not clear that deleteNullable doesn't apply to "object" logic above

sebastiantorrescl commented:
Created the .jar from the output. :D
